internal/services/cognitive/cognitive_account_connection_resource.go

```diff
@@ -0,0 +1,544 @@
+				Key: pointer.To(model.ApiKey),
+			},
+		}, nil
+
+	case accountconnectionresource.ConnectionAuthTypeOAuthTwo:
+		if len(model.OAuth2) == 0 {
+			return nil, errors.New("when `auth_type` is `OAuth2`, `oauth2` block must be specified")
+		}
+
+		return accountconnectionresource.OAuth2AuthTypeConnectionProperties{
+			AuthType: authType,
+			Category: pointer.ToEnum[accountconnectionresource.ConnectionCategory](model.Category),
+			Metadata: pointer.To(model.Metadata),
+			Target:   pointer.To(model.Target),
+			Credentials: &accountconnectionresource.ConnectionOAuth2{
+				AuthURL:        pointer.To(model.OAuth2[0].AuthURL),
+				ClientId:       pointer.To(model.OAuth2[0].ClientId),
+				ClientSecret:   pointer.To(model.OAuth2[0].ClientSecret),
+				DeveloperToken: pointer.To(model.OAuth2[0].DeveloperToken),
+				Password:       pointer.To(model.OAuth2[0].Password),
+				RefreshToken:   pointer.To(model.OAuth2[0].RefreshToken),
+				TenantId:       pointer.To(model.OAuth2[0].TenantId),
+				Username:       pointer.To(model.OAuth2[0].Username),
+			},
+		}, nil
+
+	case accountconnectionresource.ConnectionAuthTypeCustomKeys:
+		if len(model.CustomKeys) == 0 {
+			return nil, errors.New("when `auth_type` is `CustomKeys`, `custom_keys` must be specified")
+		}
+
+		return accountconnectionresource.CustomKeysConnectionProperties{
+			AuthType: authType,
+			Category: pointer.ToEnum[accountconnectionresource.ConnectionCategory](model.Category),
+			Metadata: pointer.To(model.Metadata),
+			Target:   pointer.To(model.Target),
+			Credentials: &accountconnectionresource.CustomKeys{
+				Keys: &model.CustomKeys,
+			},
+		}, nil
+
+	case accountconnectionresource.ConnectionAuthTypeAAD:
+		if model.ApiKey != "" || len(model.OAuth2) > 0 || len(model.CustomKeys) > 0 {
+			return nil, errors.New("when `auth_type` is `AAD`, no other auth configuration blocks should be specified")
+		}
+
+		return accountconnectionresource.AADAuthTypeConnectionProperties{
+			AuthType: authType,
+			Category: pointer.ToEnum[accountconnectionresource.ConnectionCategory](model.Category),
+			Metadata: pointer.To(model.Metadata),
+			Target:   pointer.To(model.Target),
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("unsupported `auth_type`: %q", model.AuthType)
+	}
+}
+
+func expandConnectionPropertiesForUpdate(model CognitiveAccountConnectionModel, d *pluginsdk.ResourceData) (accountconnectionresource.ConnectionPropertiesV2, error) {
+	// The auth_type will be used to decide which properties need updating, so we do not check hasChange for auth_type.
+	switch authType := accountconnectionresource.ConnectionAuthType(model.AuthType); authType {
+	case accountconnectionresource.ConnectionAuthTypeApiKey:
+		props := accountconnectionresource.ApiKeyAuthConnectionProperties{
+			AuthType: authType,
+		}
+
+		if d.HasChange("api_key") {
+			if model.ApiKey == "" {
+				return nil, errors.New("when `auth_type` is `ApiKey`, `api_key` must be specified")
```

Comment:
could you move it to `CustomizeDiff` function so that it can be validated in `plan` stage?

Reply:
This cannot be validated in `CustomizeDiff` because it may refer to a property from another resource, as explained in https://developer.hashicorp.com/terraform/plugin/sdkv2/resources/customizing-differences.

> NOTE: `CustomizeDiff` does not currently support computed values from other resource attributes that are only known after apply.